### ISSUE TITLE
Explicit support for HTML and fbt white space rules

### ISSUE THREAD

**Issue by karl** [prettier/prettier#2279]
At the moment the JSX printer supports the standard JSX white space rules and has implicit support for the Facebook translation (fbt) white space rules (see https://github.com/prettier/prettier/issues/1581#issuecomment-300975032 for more detail).

I believe that the fbt white space rules are the same as the HTML whitespace rules, where a new line counts as a single space when rendering.

Given we want to support pretty printing HTML (https://github.com/prettier/prettier/issues/2097#issuecomment-310641736) it seems like it might be useful to add explicit support in the JSX printer for two different white space rules (JSX and HTML). That way we could look at re-using the JSX printer code for HTML.

When printing JSX we would default to the JSX white space rules, but if we encounter an `fbt` tag (https://github.com/prettier/prettier/issues/2233#issuecomment-310855122) then we would switch to HTML whitespaces rules.

When printing HTML we would default to the HTML white space rules.

I'm interested in investigating this to see if it is a sensible idea.

**Comment by karl** [comment]
Explicitly supporting the HTML whitespace rules would make it easier to fix edge cases in `fbt` formatting such as the one raised in https://github.com/prettier/prettier/pull/2274#issuecomment-311195471

**Comment by azz** [comment]
Sounds great!

**Comment by karl** [comment]
@vjeux I'm doing some thinking about how best to support multiple white space rules, and wanted to double check the white space rules of the `<fbt>` tags.

Do they follow the same rules as HTML in all cases? 

Specifically do new lines inside a tag at the beginning and end get rendered as a space? 

```jsx
<fbt>
  X<b>
    Y
  </b>Z
</fbt>

// Is this output as:
//   X Y Z
// or:
// XYZ
```

I'm trying to ascertain whether we have two sets of white space rules:

1. JSX
2. HTML / fbt

or whether we have 3:

1. JSX
2: HTML
3: fbt

**Comment by vjeux** [comment]
cc @jrwats who implemented <fbt> parsing

**Comment by karl** [comment]
Hmm... I've just been looking over the scant examples `fbt` (such as https://github.com/prettier/prettier/issues/1581#issuecomment-300975032) and maybe I have the wrong end of the stick about how the translation stuff works.

I was thinking that it was tags that could contain any JSX or HTML, but perhaps the tags are actually replaced by plain text before JSX parsing actually happens? That might explain why there are different white space rules before/after the tags.

**Comment by jrwats** [comment]
@karl, yes with respect to newlines, <fbt> parsing shareses the same rules as HTML, i.e. all whitespace (including newlines) is normalized to one space.  There may be other cases (outside of spacing) that I'm not thinking of where the rules differ however...

**Comment by karl** [comment]
Having thought about this some more I'm still really struggling to know what transformations are safe to make with the Facebook translation pipeline in place.

I was initially thinking of changing the white space behaviour inside of `fbt` tags. This would mean that outside of `fbt` tags Prettier would be free to follow the standard JSX white space rules.

So all of the following would be equivalent:

```
<div>
  <fbt:param>
    First
  </fbt:param>
  ,
  <fbt:param>
    Second
  </fbt:param>
</div>
// First,Second

<div>
  <fbt:param>
    First
  </fbt:param>,
  <fbt:param>
    Second
  </fbt:param>
</div>
// First,Second
```

But this doesn't match @vjeux's example in https://github.com/prettier/prettier/issues/1581#issuecomment-300975032

In that example the white space outside of the `fbt` tag no longer seems to be following the JSX white space rules. Which has left me very confused!

Would it be possible of get a number of examples to use a test cases? Ideally containing examples of all the ways the `fbt` tags can be used and how they interact with the JSX tags around them (and possibly inside them?).

Even better would be some way to have access to the translation pipeline so I could play around with examples and see the results myself. Although I understand that is unlikely if all the infrastructure around this is internal to Facebook.

Long story short: I'd like to refactor the currently very messy JSX code, but it feels too risky to do that without knowing how it will affect the Facebook translation pipeline!

**Comment by karl** [comment]
@vjeux @jrwats pinging to see if there is a way to get more info on the `fbt` whitespace rules before I refactor the JSX code. Thanks!

**Comment by jrwats** [comment]
Sorry. I missed your June 28th message. I'll provide some examples tomorrow
pr Wednesday.

**Comment by karl** [comment]
Awesome! No rush with the examples as I'll be tackling this on and off over the next few weeks.

**Comment by karl** [comment]
@jrwats Sorry to be a pest, just want to ping in case you still wanted to provide examples!

**Comment by jrwats** [comment]
Ack. Yes my bad! Will do tomorrow

**Comment by jrwats** [comment]
OK, so I have some source examples with their corresponding unminified transpiled result.  You can see the implied space in the first argument to fbt._(...) as that's the composite string created.

[SomeModule.txt](https://github.com/prettier/prettier/files/1181486/SomeModule.txt)
[SomeResult.txt](https://github.com/prettier/prettier/files/1181492/SomeResult.txt)

I'm pretty sure you'll probably want some more examples.  Just let me know, what you'd like to see.

**Comment by karl** [comment]
That is great @jrwats! I'll take a look at those and get back to you if I have any more questions. Thanks!

**Comment by azz** [comment]
Love to get the ball rolling on HTML support again. @karl any further ideas on this one?

**Comment by karl** [comment]
Ah, I was just thinking about this the other day!

I started looking into this a couple of months ago after @jrwats posted the fbt examples. I even had a branch with some work in progress, although most of the tests were still failing.

I'd definitely like to look into this again. I'm thinking the best plan might be to write up my thoughts on how I would approach this, and maybe create a PR with my (currently broken) code. That would at least give us something to discuss.

**Comment by karl** [comment]
The basic idea I wanted to explore here was an intermediate representation format.

The idea is to have a collection of alternating children and separators (one of space, new line, or nothing).

```jsx
<div><br /><br /></div>
children = [nothing, BR, nothing, BR, nothing];

---

<div> Some text <br /> </div>
children = [space, "Some text", space, BR, space];

---

<div>
  <br />
  <br />
</div>
children = [newline, BR, newline, BR, newline];
```

From this intermediate representation we could then generate the output using one of 3 sets of rules:

* JSX
* FBT
* HTML

When playing around with this the idea seemed to work OK, although if I recall correctly there are some edge cases that make it difficult to cleanly separate the `AST -> Intermediate -> Prettier doc` stages.

**Comment by karl** [comment]
The work in progress code for this can be found here:
https://github.com/prettier/prettier/pull/2976

I wouldn't expect anyone to be able to make much sense of it as it a real hacked together prototype where most of the tests are still failing!

**Comment by karl** [comment]
Better late than never I've made some progress towards this by detecting `fbt` tags and handling them differently to other JSX tags:

https://github.com/prettier/prettier/pull/4717

**Comment by thorn0** [comment]
As Prettier has had HTML support for almost a year, I think this issue can be closed.
